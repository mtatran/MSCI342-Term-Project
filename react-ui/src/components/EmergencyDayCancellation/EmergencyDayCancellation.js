import React, { Component } from "react";
import queryString from 'query-string';
import _ from 'lodash';
import Modal from 'react-modal';

import "./EmergencyDayCancellation.scss";

const axios = require('axios').default;
var moment = require('moment');

export default class EmergencyDayCancellation extends Component {
    constructor(props) {
        super(props);
        this.state = {
            showModal: false,
            isSuccessfullyCancelled: false,
            cancelledDate: '',
        }

        // Binds methods to prevent issues with 'this' keyword.
        this.showModal = this.showModal.bind(this);
        this.closeModal = this.closeModal.bind(this);
        this.handleCancelledDateInput = this.handleCancelledDateInput.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
        this.cancelAppointments = this.cancelAppointments.bind(this);
    }

    showModal() {
        this.setState({ showModal: true })
    }

    closeModal() {
        this.setState({ showModal: false })
    }

    // Stores the latest value of the cancelled date to the state variable.
    handleCancelledDateInput(event) {
        const val = event.target.value;
        this.setState({
            cancelledDate: val
        });
    };

    // Handles form submission logic and cancels the worker's appointments for the specified date.
    async handleSubmit(e) {
        e.preventDefault();

        await this.cancelAppointments();
        this.closeModal();
    }

    // Calls the back end endpoint to cancel all the worker's appointments for the specific date.
    async cancelAppointments() {
        const cancelledDate = this.state.cancelledDate;
        const params = { accessToken: this.props.user.accessToken, workerId: this.props.user.personId, specificDate: cancelledDate };
        // Formats the date for better readability.
        const formattedCancelledDate = moment(cancelledDate).format('dddd, MMM D, YYYY');

        await axios.get(`/api/cancel-specific-day/?${queryString.stringify(params)}`)
            .then(res => {
                if (_.isNil(res.error) && res.data) { // If no error and the returned response is true.
                    // Display the appropriate alert box based on whether the appointments were successfully cancelled.
                    alert(`Success!\nAll appointments for ${formattedCancelledDate} have been cancelled!`);
                } else alert(`Error!\nAn issue occurred when trying to cancel all appointments for ${formattedCancelledDate}.\nPlease try again!`);
            });
    }

    render() {
        return (
            <div>
                <button className="cancelAppointmentBtn" onClick={this.showModal}>Cancel All Appointments for a Specific Day</button>
                {/* Only displays the modal if the user has clicked the button to show the form. */}
                <Modal isOpen={this.state.showModal}>
                    <button onClick={this.closeModal} className="closeBtn">X</button>
                    <h1 align="center">Cancel Appointments for Specific Day</h1>
                    <form onSubmit={this.handleSubmit} className="cancelDayForm">
                        <label>Select the Specific Day: </label>
                        <input
                            type="date"
                            min={new Date().toISOString().split('T')[0]}
                            required
                            className="cancelledDateInput"
                            onChange={this.handleCancelledDateInput} /> <br />
                        <input
                            className="submitBtn"
                            type="submit"
                            value="Cancel Day's Appointments" />
                    </form>
                </Modal>
            </div>
        );

    }
}
