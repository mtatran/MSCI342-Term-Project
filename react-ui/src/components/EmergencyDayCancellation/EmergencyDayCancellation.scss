@import "../Common/ColorSchema.scss";

.cancelAppointmentBtn {
    background-color: $emergency-btn-color;
    position: absolute;
    right: 0;
    padding: 5px;
    margin-right: 10px;
    font-size: $typical-font-size;   
}

.closeBtn {
    float: right;
    padding: 20px;
    font-size: $typical-font-size;
    background-color: $worker-active-color;
}

.cancelDayForm {
    display: block;
    text-align: center;
    font-size: $typical-font-size;
    margin-top: 5%;
}

.cancelledDateInput {
    text-align: center;
    font-size: $input-font-size;
}

.submitBtn {
    background-color: $emergency-btn-color;
    margin-top: 5%;
    padding: 1%;
    font-size: $typical-font-size;
}