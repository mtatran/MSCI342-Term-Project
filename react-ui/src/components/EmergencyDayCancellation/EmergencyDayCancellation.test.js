import React from "react";
import { create } from "react-test-renderer";
import { render, waitFor, cleanup, fireEvent } from '@testing-library/react';
import * as axios from 'axios';
import Modal from 'react-modal';
import ReactDOM from 'react-dom';
import TestRenderer from 'react-test-renderer';

import EmergencyDayCancellation from './EmergencyDayCancellation';

jest.mock('axios');

describe('EmergencyDayCancellation component', () => {
    afterEach(cleanup);

    beforeEach(() => {
        jest.resetModules(); // Clears any cache between tests.
    });

    test('Snapshot successfully renders', () => {
        // Arrange
        // No variables needed for this test case.

        // Act
        const component = create(<EmergencyDayCancellation />);

        // Assert
        expect(component.toJSON()).toMatchSnapshot();
    });

    test('modal not initially shown', () => {
        // Arrange
        const isInitialModalShown = false;

        // Act
        const testRenderer = TestRenderer.create(<EmergencyDayCancellation />);
        const testInstance = testRenderer.root;

        // Arrange
        expect(testInstance.findByType(Modal).props.isOpen).toBe(isInitialModalShown);
    });

    test('modal shown after appropriate button click', () => {
        // Arrange
        const isModalShownAfterClick = true;

        // Act
        const testRenderer = TestRenderer.create(<EmergencyDayCancellation />);
        const testInstance = testRenderer.root;
        ReactDOM.createPortal = node => node; // Needed to mock the portal parent for clicking.
        testInstance.findByProps({ className: "cancelAppointmentBtn" }).props.onClick();

        // Arrange
        expect(testInstance.findByType(Modal).props.isOpen).toBe(isModalShownAfterClick);
    });

    test('successful submission of the cancellation modal with a date inputted with correct user props', async () => {
        // Arrange
        const testUser = {
            userType: 'worker',
            personId: '8000000',
            accessToken: 'XcCa92ZvOnQKZsGtOKOa',
        };
        const selectedDate = '2020-12-15';
        const resp = { data: true };
        const successfulSubmissionMsg = "Success!\nAll appointments for Tuesday, Dec 15, 2020 have been cancelled!";
        axios.get.mockImplementation(() => Promise.resolve(resp));
        window.alert = jest.fn();

        // Act
        const testRenderer = TestRenderer.create(<EmergencyDayCancellation user={testUser} />);
        const testInstance = testRenderer.root;
        ReactDOM.createPortal = node => node; // Needed to mock the portal parent for clicking.
        testInstance.findByProps({ className: "cancelAppointmentBtn" }).props.onClick(); // Opens the modal.
        testInstance.findByProps({ className: "cancelledDateInput" }).props.onChange({ target: { value: selectedDate } }); // Simulates user input of date.
        await testInstance.findByProps({ className: "cancelDayForm" }).props.onSubmit({ preventDefault: () => { } }); // Submits the form.

        // Arrange
        expect(window.alert).toHaveBeenCalledWith(successfulSubmissionMsg);
    });

    test('failed submission of the cancellation modal with a date inputted with correct user props', async () => {
        // Arrange
        const testUser = {
            userType: 'worker',
            personId: '8000000',
            accessToken: 'XcCa92ZvOnQKZsGtOKOa',
        };
        const selectedDate = '2020-12-15';
        const resp = { data: false };
        const failedMsg = "Error!\nAn issue occurred when trying to cancel all appointments for Tuesday, Dec 15, 2020.\nPlease try again!";
        axios.get.mockImplementation(() => Promise.resolve(resp));
        window.alert = jest.fn();

        // Act
        const testRenderer = TestRenderer.create(<EmergencyDayCancellation user={testUser} />);
        const testInstance = testRenderer.root;
        ReactDOM.createPortal = node => node; // Needed to mock the portal parent for clicking.
        testInstance.findByProps({ className: "cancelAppointmentBtn" }).props.onClick(); // Opens the modal.
        testInstance.findByProps({ className: "cancelledDateInput" }).props.onChange({ target: { value: selectedDate } }); // Simulates user input of date.
        await testInstance.findByProps({ className: "cancelDayForm" }).props.onSubmit({ preventDefault: () => { } }); // Submits the form.

        // Arrange
        expect(window.alert).toHaveBeenCalledWith(failedMsg);
    });

    test('submission of the cancellation modal with no user props but a date is inputted', async () => {
        // Arrange
        const selectedDate = '2020-12-15';
        var isThrown = false;
        const resp = { data: false };
        axios.get.mockImplementation(() => Promise.resolve(resp));

        // Act
        try {
            const testRenderer = TestRenderer.create(<EmergencyDayCancellation />);
            const testInstance = testRenderer.root;
            ReactDOM.createPortal = node => node; // Needed to mock the portal parent for clicking.
            testInstance.findByProps({ className: "cancelAppointmentBtn" }).props.onClick(); // Opens the modal.
            testInstance.findByProps({ className: "cancelledDateInput" }).props.onChange({ target: { value: selectedDate } }); // Simulates user input of date.
            await testInstance.findByProps({ className: "cancelDayForm" }).props.onSubmit({ preventDefault: () => { } }); // Submits the form.
        } catch (err) {
            isThrown = true;
        }

        // Arrange
        expect(isThrown).toBe(true);
    });

    test('submission of the cancellation modal with invalid date inputted and correct user props', async () => {
        const testUser = {
            userType: 'worker',
            personId: '8000000',
            accessToken: 'XcCa92ZvOnQKZsGtOKOa',
        };
        const failedMsg = "Error!\nAn issue occurred when trying to cancel all appointments for Invalid date.\nPlease try again!";
        window.alert = jest.fn();

        // Act
        const testRenderer = TestRenderer.create(<EmergencyDayCancellation user={testUser} />);
        const testInstance = testRenderer.root;
        ReactDOM.createPortal = node => node; // Needed to mock the portal parent for clicking.
        testInstance.findByProps({ className: "cancelAppointmentBtn" }).props.onClick(); // Opens the modal.
        testInstance.findByProps({ className: "cancelledDateInput" }).props.onChange({ target: { value: "YYYY-MM-DD" } }); // Empty value for the date.
        await testInstance.findByProps({ className: "cancelDayForm" }).props.onSubmit({ preventDefault: () => { } }); // Submits the form.

        // Arrange
        expect(window.alert).toHaveBeenCalledWith(failedMsg);
    });

    test('submission of the cancellation modal with no date inputted and correct user props', async () => {
        const testUser = {
            userType: 'worker',
            personId: '8000000',
            accessToken: 'XcCa92ZvOnQKZsGtOKOa',
        };
        const failedMsg = "Error!\nAn issue occurred when trying to cancel all appointments for Invalid date.\nPlease try again!";
        window.alert = jest.fn();

        // Act
        const testRenderer = TestRenderer.create(<EmergencyDayCancellation user={testUser} />);
        const testInstance = testRenderer.root;
        ReactDOM.createPortal = node => node; // Needed to mock the portal parent for clicking.
        testInstance.findByProps({ className: "cancelAppointmentBtn" }).props.onClick(); // Opens the modal.
        await testInstance.findByProps({ className: "cancelDayForm" }).props.onSubmit({ preventDefault: () => { } }); // Submits the form.

        // Arrange
        expect(window.alert).toHaveBeenCalledWith(failedMsg);
    });

});
