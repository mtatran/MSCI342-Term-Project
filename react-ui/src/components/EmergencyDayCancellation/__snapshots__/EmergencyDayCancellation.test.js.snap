// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmergencyDayCancellation component Snapshot successfully renders 1`] = `
<div>
  <button
    className="cancelAppointmentBtn"
    onClick={[Function]}
  >
    Cancel All Appointments for a Specific Day
  </button>
</div>
`;
